cpp/src/filling/fill.cu

```diff
@@ -0,0 +1,158 @@
+
+#include <memory>
+
+namespace {
+
+struct scalar_factory {
+  gdf_scalar value;
+
+  template <typename T>
+  struct scalar {
+    T value;
+    bool is_valid;
+
+    __device__
+    T data(cudf::size_type index) { return value; }
+
+    __device__
+    bool valid(cudf::size_type index) { return is_valid; }
+  };
+
+  template <typename T>
+  scalar<T> make() {
+    T val{}; // Safe type pun, compiler should optimize away the memcpy
+    memcpy(&val, &value.data, sizeof(T));
+    return scalar<T>{val, value.is_valid};
+  }
+};
+
+}  // namespace
+
+namespace cudf {
+namespace experimental {
+
+namespace detail {
+
+void fill(mutable_column_view& destination,
+          size_type begin,
+          size_type end,
+          gdf_scalar const& value,
+          cudaStream_t stream) {
+  CUDF_EXPECTS(cudf::is_fixed_width(destination.type()) == true,
+               "In-place fill does not support variable-sized types.");
+  CUDF_EXPECTS((begin >= 0) &&
+               (begin <= end) &&
+               (begin < destination.size()) &&
+               (end <= destination.size()),
+               "Range is out of bounds.");
+  CUDF_EXPECTS((destination.nullable() == true) || (value.is_valid == true),
+               "destination should be nullable or value should be non-null.");
+  // The line below should be enabled once we switch from gdf_scalar to
+  // cudf::scalar.
+  // CUDF_EXPECTS(destination.type() == value.type(), "Data type mismatch.");
+
+  if (end != begin) {  // otherwise no-op
+    copy_range(destination, scalar_factory{value}, begin, end, stream);
+  }
+
+  return;
+}
+
+std::unique_ptr<column> fill(column_view const& input,
+                             size_type begin,
+                             size_type end,
+                             gdf_scalar const& value,
+                             cudaStream_t stream,
+                             rmm::mr::device_memory_resource* mr) {
+  CUDF_EXPECTS(cudf::is_fixed_width(input.type()) == true,
+               "Variable-sized types are not supported yet.");
+  CUDF_EXPECTS((begin >= 0) &&
+               (begin <= end) &&
+               (begin < input.size()) &&
+               (end <= input.size()),
+               "Range is out of bounds.");
+  // The line below should be enabled once we switch from gdf_scalar to
+  // cudf::scalar.
+  // CUDF_EXPECTS(input.type() == value.type(), "Data type mismatch.");
+
+  auto state = mask_state{UNALLOCATED};
+  if ((input.nullable() == true) || (value.is_valid == false)) {
+    state = UNINITIALIZED;
+  }
+
+  auto ret = std::unique_ptr<column>{nullptr};
+
+  if (cudf::is_numeric(input.type()) == true) {
+    ret = make_numeric_column(input.type(), input.size(), state, stream, mr);
+  }
+  else if (cudf::is_timestamp(input.type()) == true) {
+    ret = make_timestamp_column(input.type(), input.size(), state, stream, mr);
+  }
+  else {
+    CUDF_FAIL("Unimplemented.");
+  }
+
+  auto destination = ret->mutable_view();
+  if (begin > 0) {
```

Comment:
Likewise, just do:
```
auto ret = std::make_unique<column>{input}
// fill in [begin,end) with `value`. 
```

Reply:
@jrhemstad What should I do if _input_ is not nullable but _value_ is invalid (filling with null values)? In this case, I need to allocate memory for null masks even when _input_ is non-nullable, and I don't see an appropriate constructor for this.

Reply:
Hm, this isn't the cleanest, but it works:
```
column copy{input}; // non-nullable copy
rmm::device_buffer mask = create_null_mask(input.size(), ALL_VALID);
auto contents = copy.release();
column c(input.type(), input.size(), std::move(contents.data), std::move(mask), input.null_count());
```

Reply:
This looks more complex than the original code (considering that we're dropping the original code preferring a simpler solution).

I can create

```
column::column(column const& other, mask_state state);
```
And this constructor makes sense if _other_ is not nullable, but does not make very much sense for columns with null masks; say, if column _a_ has null elements and _state_ is set to ALL_VALID or UNALLOCATED, the resulting column will have garbage values as valid elements.

What do you think about this? Just disallow calling this constructor with columns with null values (so, throw an exception) or allow garbage values becoming valid elements assuming that users actually asked for this?

Reply:
Another option is to create something like
```
std::unique_ptr<column>
duplicate(
  column_view const& input,
  mask_allocation_policy mask_alloc,
  rmm::mr::device_memory_resource *mr);
```
in copying.hpp (and do something similar to what you have suggested inside the duplicate function), but the same problem (garbage values becoming valid elements) exists if _input_ has null values but mask_alloc is set to NEVER.